src/clean_air/util/file_converter.py

```diff
@@ -38,11 +41,73 @@ def generate_dataframe(filepath):
         # that into a pandas dataframe.
         temp_dataset = xr.open_dataset(filepath)
         temp_dataframe = temp_dataset.to_dataframe()
+    elif filepath.endswith(".csv"):
+        # Not only do we need to read this format into a dataframe, but we
+        # also need to reformat it into acceptable standards before passing
+        # it on to saving functions.
+        temp_dataframe = csv_reformatter(filepath)
     else:
         raise ValueError("No reader configured yet for this input format.")
     return pd.DataFrame(temp_dataframe)
 
 
+def csv_reformatter(filepath):
+    """This function uses a template to reformat known-source CSV files into a
+    fixed format that we can then very easily read into a pandas dataframe
+    (ready for conversion to netcdf files)."""
+    # rules for csv format:
+    # - header line (column names) must be on row 4 of file (top row is 0)
+    # - column names must be standard python strings.
+
+    # This is a dictionary containing names to look for and what to convert
+    # them into.
+    cap_converters = dict(
+        {'PM<sub>10</sub> particulate matter (Hourly measured)': 'PM10',
+         'Non-volatile PM<sub>10</sub> (Hourly measured)': 'Non-volatilePM10',
+         'Non-volatile PM<sub>2.5</sub> (Hourly measured)': 'Non-volatilePM2p5',
+         'PM<sub>2.5</sub> particulate matter (Hourly measured)': 'PM2p5',
+         'Volatile PM<sub>10</sub> (Hourly measured)': 'Volatile PM10',
+         'Volatile PM<sub>2.5</sub> (Hourly measured)': 'Volatile PM2p5',
+         'Ozone': 'O3',
+         'Nitric Oxide': 'NO',
+         'Nitrogen Dioxide': 'NO2',
+         'Sulphur Dioxide': 'SO2',
+         'Carbon Monoxide': 'CO',
+         'Date': 'Date',
+         'time': 'time'})
+
+    # Now that we have some things to look for we can read the header row of
+    # the file to pull out the applicable column names:
+    data_names = pd.read_csv(filepath, nrows=1, skiprows=4, delimiter=',')
+
+    # We need to narrow down our own list of column names by matching them
+    # to those we are looking for:
+    good_names = []
+    bad_names = []
+    for n, name in enumerate(data_names.columns):
+        if name in (cap_converters.keys() or cap_converters.values()):
+            good_names.append(name)
+        else:
+            bad_names.append(name)
+
+    # And then we need to produce a shorter conversion list containing only
+    # the conversions required in this file:
+    conv_list = {}
+    for k, v in cap_converters.items():
+        if (k or v) in good_names:
```

Comment:
I'm puzzled why we're checking the values. Surely we only expect field names to appear in the keys?

Reply:
This was part of the extension to less formatted files; if we had some data files from other sources they may not use the same header names as AURN do, so I thought it might be useful to check for names in the format we are using as well, just in case.

Still undecided about whether this is actually useful or not.